src/szk.rs

```diff
@@ -0,0 +1,444 @@
+            } else {
+                (None, None, Vec::<T::Field>::new())
+            };
+
+        let prove_rand = self.derive_prove_rands(&prove_rand_seed);
+        let mut leader_proofs_share =
+            self.typ
+                .prove(encoded_measurement, &prove_rand, &joint_rand)?;
+
+        // Generate the proof shares.
+        for (x, y) in leader_proofs_share
+            .iter_mut()
+            .zip(self.derive_helper_proofs_share(&helper_seed))
+        {
+            *x -= y;
+        }
+
+        // Prep the output messages.
+        let leader_proof_share = SzkProofShare::Leader {
+            uncompressed_proof_share: leader_proofs_share,
+            blind_and_joint_rand_part: blind_and_joint_rand_part_for_leader,
+        };
+        let helper_proof_share = SzkProofShare::Helper {
+            proof_share_seed_and_blind: helper_seed.clone(),
+            joint_rand_part: helper_joint_rand_part,
+        };
+        Ok([leader_proof_share, helper_proof_share])
+    }
+
+    fn query(
+        &self,
+        input_share: &[T::Field],
+        proof: SzkProofShare<SEED_SIZE, T::Field>,
+        verify_key: &[u8; SEED_SIZE],
+        nonce: &[u8; 16],
+    ) -> Result<SzkQueryShares<SEED_SIZE, T::Field>, SzkError> {
+        let query_rand = self.derive_query_rands(verify_key, nonce);
+
+        let agg_id = match proof {
+            SzkProofShare::Leader { .. } => 0,
+            SzkProofShare::Helper { .. } => 1,
+        };
+
+        let proof_share = match proof {
+            SzkProofShare::Leader {
+                ref uncompressed_proof_share,
+                ..
+            } => Cow::Borrowed(uncompressed_proof_share),
+            SzkProofShare::Helper {
+                ref proof_share_seed_and_blind,
+                ..
+            } => Cow::Owned(self.derive_helper_proofs_share(proof_share_seed_and_blind)),
+        };
+
+        let (joint_rand, jr_seed_and_part) = if self.typ.joint_rand_len() > 0 {
+            let (seed, other_joint_rand_part) = match proof {
+                SzkProofShare::Leader {
+                    uncompressed_proof_share: _,
+                    blind_and_joint_rand_part,
+                } => blind_and_joint_rand_part.expect("jr_len > 0 so there is joint_randomness"),
+                SzkProofShare::Helper {
+                    proof_share_seed_and_blind,
+                    joint_rand_part,
+                } => (
+                    proof_share_seed_and_blind,
+                    joint_rand_part.expect("jr_len > 0 so there is joint randomness"),
+                ),
+            };
+            let own_joint_rand_part = self.derive_joint_rand_part(&seed, input_share, nonce)?;
+            let (leader_joint_rand_part, helper_joint_rand_part) = match agg_id {
+                0 => (&own_joint_rand_part, &other_joint_rand_part),
+                1 => (&other_joint_rand_part, &own_joint_rand_part),
+                _ => return Err(SzkError::Query("bad agg ID".to_string())),
```

Comment:
nit: This is unreachable since agg_id is set to zero or one above. We could tidy this up by replacing `agg_id` with `is_leader: bool`.